# PlasChain

PlasChain: an algorithm for improving long plasmid reconstruction from metagenome assemblies.

- [Installation](#installation)
  * [With Conda](#with-conda)
      - [From yaml file](#with-environment-yaml-file)
- [Database](#database)
- [Testing your PlasChain installation](#testing-your-plaschain-installation)
- [Basic Usage](#basic-usage)
- [Main output files](#main-output-files)
- [Advanced usage](#advanced-usage)
  * [Plasmid-specific genes](#plasmid-specific-genes)
- [Simulation script](#simulation-script)
  * [Reference Preparation](#Reference-Preparation)
  * [Basic Usage](#Basic-script-Usage)

### With Conda

#### With environment yaml file

Create and activate the conda environment:
```
conda env create -f environment.yaml
conda activate PlasChain
```




## Database
PlasChain requires the plasmid-specific genes (PSGs) stored in the data directory to annotate the assembly graph.
Platon depends on a custom database which can be download here.
 [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.4066768.svg)](https://doi.org/10.5281/zenodo.4066768)

```bash
$ wget https://zenodo.org/record/4066768/files/db.tar.gz
$ tar -xzf db.tar.gz
$ rm db.tar.gz
```
## Testing your PlasChain installation
Once you have completed the above you can run 
```
cd PlasChain
chmod u+x ./run_install_test.sh
./run_install_test.sh <path_to_platon_db>
```
from the outermost PlashChain directory to test your installation.

Since the database required by Platon needs to be downloaded manually, you must provide the path to the database when running the test script.

It may take a few minutes, depending on your device's performance.


## Basic Usage
To run the PlasChain pipeline: 
```
python plaschain.py -g <fastg graph> -o <output directory> [-k <max k value>] -r1 <reads 1> -r2 <reads 2> -path <contig path file> -db <platon db> [-p <num processes>]
```
If a BAM alignment file of the reads to the assembly graph already exists, then use the following command to avoid re-running the alignment:
```
python plaschain.py -g <fastg graph> -o <output directory> [-k <max k value>] -b <BAM file> -path <contig path file> -db <platon db>  [-p <num processes>]
```
The common command line options are:

`-g/--graph`: : Assembly graph fastg file.

`-o/--output_dir`: Output directory.

`-k/max_k`: Maximum k value used by the assembler. Default: 77.

`-p/--num_processes`: Number of processes to use. Default: 16.

`-r1/--reads1`: Paired-end reads file 1.

`-r2/--reads2`: Paired-end reads file 2.

`-b/--bam`: BAM alignment file aligning reads to graph nodes `-b` and `-r1`,`-r2` are mutually exclusive.

`-path/--contig_path`: Contig path generated by metaSPAdes.

`-db/--platon_db`: DB required by Platon.

## Main output files
The output files are written in the directory specified by the user:

The file `<prefix>.confident_cycs.fasta` contains the assembled plasmid sequences. The `<prefix> `is derived from the input filename provided via the `-g` parameter, with the `.fastg` extension removed.

There are some files in `intermediate_files` directory 
  
- `<prefix>.cycs.fasta` is a fasta file of **all** cyclic paths that were considered as potential plasmids before filtering to create the subset that is output.
  
- `<prefix>.cycs.paths.txt` describes the graph nodes that constitute each cyclic path, corresponding to the sequences in `<prefix>.cycs.fasta.`
  
- `<prefix>.classified_cycs.fasta` is a fasta file of cyclic paths with score > 0.5 in `<prefix>.cycs.fasta`.
  
- `<prefix>.gene_filtered_cycs.fasta` is a fasta file of cyclic paths with PSG hit in `<prefix>.cycs.fasta`.
  
- `<prefix>.self_loops_fasta` is a fasta file of self-loop paths in `<prefix>.cycs.fasta`.
  
- `<prefix>.contigs.fasta` is a fasta file of contig path that meets the length constriction.
  
- `<prefix>.contigs.score.fasta` reports the score of contig path in `<prefix>.contigs.fasta`.
  
- `<prefix>.cycs.plasmid.fasta` is a fasta file of cyclic paths predicted as plasmids by platon.
  
- `reads_pe_primary.sort.bam(.bai)` is the alignment file for the reads to the assembly graph. It can be re-used with the `-b` option if PlasChain is re-run on the same sample.

Under the `logs` subdirectory, the file `plaschain.log` contains all information about the PlasChain run. Other log files in the `logs` subdirectory may be helpful if there is an error or failure in one of the stages that runs BLAST, BWA, or PlasClass.

## Advanced usage
More advanced command line options allow for different stages in the PlasChain pipeline to be modified:

`-sc/--use_scores`: Flag to determine whether to use plasmid scores. Use `False` to turn off plasmid score use. Default: `True`.

`-gh/--use_gene_hits`: Flag to determine whether to use plasmid specific genes. Use `False` to turn off plasmid gene use. Default: `True`.

`-pc/--plasclass`: PlasClass score file. If PlasClass classification of the assembly graph nodes has already been performed, provide the name of the PlasClass output file. (For example: the `intermediate_files/plasclass.out` file from a previous run of PlasChain).

In addition, all of the different thresholds used in the algorithm can be changed by the user:

`-m/--max_CV`: Maximum allowed coefficient of variation for coverage. Default: 0.5.

`-l/--min_length`: Minimum allowed length for potential plasmid. Default: 1000.

`-clft/--classification_thresh`: Threshold for classifying a potential plasmid as a plasmid. Default: 0.5.

`-gm/--gene_match_thresh`: Threshold for % identity and fraction of length covered to determine plasmid gene matches. Default: 0.75.

`-sls/selfloop_score_thresh`: Threshold plasmid score above which a self-loop is considered a potential plasmid. Default: 0.9.

`-slm/--selfloop_mate_thresh`: Threshold fraction of off-loop mate-pairs, below which a self-loop is considered a potential plasmid. Default:0.1.

`-cst/--chromosome_score_thresh`: Threshold score, below which a long node is considered a chromosome node. Default: 0.2.

`-clt/--chromosome_length_thresh`: Threshold length, above which a low scoring node is considered a chromosome node. Default: 10000.

`-pst/--plasmid_score_thresh`: Threshold score, above which a long node is considered a plasmid node. Default: 0.9.

`-plt/--plasmid_length_thresh`: Threshold length, above which a high scoring node is considered a plasmid node. Default: 10000.

`-min_clen/--min_contig_path_len`: Minimum allowed number of nodes in a valid contig path. Default: 4.

Instead of inputting all of these options on the command-line before each run of PlasChain, the user can change them in the file `params.json`. Set each variable in this file to the desired value and it will be used in PlasChain. Any value passed as a command-line parameter will override the values set in this file.

## Simulation Script
PlasChain uses InSilicoSeq with the HiSeq model (default read length = 126bp) to simulate the pair-end reads.Thus it is required by install [InSilicoSeq](https://github.com/HadrienG/InSilicoSeq) before run the script.
```
conda install -c bioconda insilicoseq=1.5.4 biopython=1.78
```

To test the script: 
```
cd PlasChain/simulate_script
chmod u+x ./*.sh ./*.awk
./run.sh 1 1 100000
```

### Reference Preparation
- You can download the reference plasmids provided by SCAPP through `Simulation.tsv` and locate them in
  `PlasChain/simulate_script/src/all_bacteria`.
- You can download some short plasmids (<10kbp) from NCBI as supplementary and locate them in
  `PlasChain/simulate_script/src/short_plasmid`.
- You can download some reference bacteria genomes from RefSeq when you want to simulate larger dataset and locate them in
  `PlasChain/simulate_script/src/refseq`.
### Basic Script Usage
After reference preparation, you can run the script: 
```
cd PlasChain/simulate_script
chmod u+x ./*.sh ./*.awk
./run.sh <num of bacteria> <num of supplementary short plasmid> <num of reads>"
```
Then you can get the simulated pair-end reads in `PlasChain/simulate_script/simulation`.

There are some other files may be useful:
- `PlasChain/simulate_script/chromosome_size.txt` reportes the length of reference chromosome.
- `PlasChain/simulate_script/plasmid_size.txt` reportes the length of reference plasmid.
- `PlasChain/simulate_script/simulation_info.txt` reportes the overall infomation about the simulation.
- `PlasChain/simulate_script/selected_fna.txt` reportes the selected reference genome.
- `PlasChain/simulate_script/association.txt` provides the mapping between each reference plasmid and its corresponding host chromosome; a null value indicates that the plasmid was randomly selected as a supplementary sequence (i.e., not derived from any real host chromosome).
- `PlasChain/simulate_script/inter/ua.txt` provides the mapping between each reference plasmid and its assigned host chromosome; supplementary plasmids are randomly attached to existing host chromosomes.

### Reproducing the Paper's Results

The `simulation_data` folder contains all necessary inputs: bacterial genomes, randomly introduced short plasmids, and chromosomal abundance data.

Steps:

1. Split the bacterial whole genomes into separate chromosome and plasmid files.
2. Simulate the chromosomal component using iss based on the provided abundance.
3. Randomly associate the selected short plasmids with the existing chromosomes to define host-plasmid pairs.
4. Compute plasmid copy numbers based on the host chromosome copy numbers (see Supplementary Material for formulas) to determine plasmid abundance and read counts.
5. Simulate the plasmid component using iss.
6. Concatenate the chromosomal and plasmid reads.
